src/libraries/System.Net.Quic/src/System/Net/Quic/QuicListener.PendingConnection.cs

```diff
@@ -0,0 +1,61 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Net.Security;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Quic;
+
+public sealed partial class QuicListener
+{
+    private sealed class PendingConnection : IAsyncDisposable
+    {
+        private static readonly TimeSpan s_handshakeTimeout = TimeSpan.FromSeconds(10);
+
+        private readonly TaskCompletionSource<QuicConnection?> _finishHandshakeTask;
+        private readonly CancellationTokenSource _cancellationTokenSource;
+
+        public PendingConnection()
+        {
+            _finishHandshakeTask = new TaskCompletionSource<QuicConnection?>(TaskCreationOptions.RunContinuationsAsynchronously);
+            _cancellationTokenSource = new CancellationTokenSource();
+        }
+
+        public async void StartHandshake(QuicConnection connection, SslClientHelloInfo clientHello, Func<QuicConnection, SslClientHelloInfo, CancellationToken, ValueTask<QuicServerConnectionOptions>> connectionOptionsCallback)
```

Comment:
I did understand that you don't ever want to await it (though I missed it's `async void`), what was strange to me is that method named "Start..." waits until finish 😅 Maybe adding a comment highlighting that it is called Start.. because it is async void so it is never awaited will be better 😅 

P.S.: I expected to see something like this https://github.com/dotnet/runtime/blob/e55c908229e36f99a52745d4ee85316a0e8bb6a2/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L502-L506